Custom imports handling in configuration file

ScottDelbecq opened the pull request:
This PR proposes a way to specify custom imports in the configuration file and resolves #413.
It assumes the user has in his Python environment the necessary packages.

christophe-david commented on src/fastoad/io/configuration/configuration.py:
```diff
@@ -66,6 +67,8 @@ class for configuring an OpenMDAO problem from a configuration file
 
     def __init__(self, conf_file_path: Union[str, PathLike] = None):
         self._serializer: _IDictSerializer = _YAMLSerializer()
+        # Store imported classes
```
This kind of comment suggests an action.
(and an empty line would be good for sake of clarity and consistency)
```suggestion

        # for storing imported classes
```

ScottDelbecq replied:
Corrected in [b5e6c07](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/b5e6c07c475e0b4c6a90b747cb483b50b17db6f5)

christophe-david commented on src/fastoad/io/configuration/tests/data/conf_with_imports.yml:
```diff
@@ -0,0 +1,29 @@
+
+
+  imports:
+      my_driver_1: MyDriver1
+      my_driver_2: MyDriver2
```
I would suggest something like that (with the proper modules added in tests/data/...):
```suggestion
      fastoad.io.configuration.tests.data.to_be_imported.my_driver_1: MyDriver1
      "fastoad.io.configuration.tests.data.to_be_imported.my_driver_2": MyDriver2
```
It would test "complex paths" and would avoid all the boilerplate at the beginning of your test.

ScottDelbecq replied:
Corrected in [9d23586](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/9d23586af2929d71c02c34f403f2546d075f6dc7)

christophe-david commented on src/fastoad/io/configuration/tests/test_configuration.py:
```diff
@@ -336,3 +337,35 @@ def test_set_optimization_definition(cleanup):
+
+
+def test_imports_handling():
+    # Mock the importlib.import_module to return a dummy class
+    class DummyClass:
+        pass
+
+    # Create a dummy module and add it to sys.modules
+    dummy_module_spec = importlib.util.spec_from_loader("my_driver_1", loader=None)
+    dummy_module = importlib.util.module_from_spec(dummy_module_spec)
+    dummy_module.MyDriver1 = DummyClass
+    import sys
+
+    sys.modules["my_driver_1"] = dummy_module
+
+    dummy_module_spec = importlib.util.spec_from_loader("my_driver_2", loader=None)
+    dummy_module = importlib.util.module_from_spec(dummy_module_spec)
+    dummy_module.MyDriver2 = DummyClass
+    sys.modules["my_driver_2"] = dummy_module
```
With my suggestion above, you can remove all that part.

ScottDelbecq replied:
Corrected in [9d23586](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/9d23586af2929d71c02c34f403f2546d075f6dc7)

christophe-david commented on src/fastoad/io/configuration/tests/test_configuration.py:
```diff
@@ -336,3 +337,35 @@ def test_set_optimization_definition(cleanup):
+    # Check that the classes are correctly imported and stored
+    assert "MyDriver1" in conf._imported_classes
+    assert "MyDriver2" in conf._imported_classes
+    assert conf._imported_classes["MyDriver1"] == DummyClass
+    assert conf._imported_classes["MyDriver2"] == DummyClass
+
+    # Check that _om_eval can use the imported classes
+    result = conf._om_eval("MyDriver1()")
+    assert isinstance(result, DummyClass)
```
Still based on my suggestion, I would recommend delaying the imports we need for checking results.
I find it important to check that FASTOADProblemConfigurator behaves correctly even if Python does not know about the modules to import.
```suggestion

    result = conf._om_eval("MyDriver1()")

    from .data.to_be_imported.my_driver_1 import MyDriver1
    from .data.to_be_imported.my_driver_2 import MyDriver2

    assert conf._imported_classes["MyDriver1"] == MyDriver1
    assert conf._imported_classes["MyDriver2"] == MyDriver2

    # Check that _om_eval can use the imported classes
    assert isinstance(result, MyDriver1)

```

ScottDelbecq replied:
Corrected in [9d23586](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/9d23586af2929d71c02c34f403f2546d075f6dc7)

christophe-david reviewed (changes requested):
Nice addition. I am ok with the implementation. I have suggestions for the test, though.

And Sphinx doc should also be updated.

You could also add another field in the configuration file, to allow the user to add paths in `sys.path`. It would allow the usage of local code.

I also wonder about error management. You will have standard Python errors if the import failed because the `imports` section is incorrectly filled. But is it worth having a crash if, for instance, the bugged import is not used? I am still undecided about that...

ScottDelbecq commented on src/fastoad/io/configuration/configuration.py:
```diff
@@ -419,19 +428,20 @@ def _add_design_vars(self, model, auto_scaling):
 
+    def _om_eval(self, string_to_eval: str):
+        """
+        Evaluates strings that assume `import openmdao.api as om` is done.
```
Corrected in [a3ce08f](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/a3ce08fd9d4978d71d9625d4cedf42adde3dd98a)

ScottDelbecq commented on docs/documentation/usage.rst:
```diff
@@ -30,6 +30,12 @@ A quick tutorial for YAML (among many ones) is available
+        my_driver_1: MyDriver1
+        utils.my_drivers.my_driver_2: MyDriver2
+        utils.my_solvers.my_favourite_solver: MyFavouriteSolver
+
```
Corrected in [1ff38e9](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/1ff38e98a5b4b32b725a2d91482be5d08ba87cc9)

ScottDelbecq commented on src/fastoad/io/configuration/tests/test_configuration.py:
```diff
@@ -336,3 +336,29 @@ def test_set_optimization_definition(cleanup):
+    result1 = conf._om_eval("MyDriver1()")
+    result2 = conf._om_eval("MyDriver2()")
+
+    from .data.to_be_imported.my_driver_1 import MyDriver1
```
Corrected in [e96c9af](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/e96c9af2c0fb02e24e32359f4296070a17230cfb)

ScottDelbecq commented:
> Nice addition. I am ok with the implementation. I have suggestions for the test, though.
> 
> And Sphinx doc should also be updated.
> 
> You could also add another field in the configuration file, to allow the user to add paths in `sys.path`. It would allow the usage of local code.
> 
> I also wonder about error management. You will have standard Python errors if the import failed because the `imports` section is incorrectly filled. But is it worth having a crash if, for instance, the bugged import is not used? I am still undecided about that...

I agree. I added the possibility to specify `sys.path` to be included and also I added a way to do a more advanced setting for drivers:

```yaml
driver:
  instance: om.ScipyOptimizeDriver(optimizer='COBYLA')
  options:
    maxiter: 100
    tol: 1e-2
  advanced_options:
    print_live_convergence: true
```

That will do the equivalent of:

```python
driver = om.ScipyOptimizeDriver(optimizer='COBYLA')
driver.options['maxiter'] = 100
driver.options['tol'] = 1e-2
driver.advanced_options['print_live_convergence'] = True
```

christophe-david commented on src/fastoad/io/configuration/tests/test_configuration.py:
```diff
@@ -336,3 +341,108 @@ def test_set_optimization_definition(cleanup):
+    result1 = conf._om_eval("MyDriver1()")
+    result2 = conf._om_eval("MyDriver2()")
+
+    # Imports are done here to verify that the classes are correctly imported
```
Sorry, I'm still not satisfied 😅 

I think a future reader will still not understand why these imports are not done at the beginning of the file. The point is we want to be sure the interpreter does not know these paths when the code above is run.

ScottDelbecq replied:
Corrected in [bad1ec0](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/bad1ec0e8606d5d30682a6bb9d23991a2dd6386c)

christophe-david commented on src/fastoad/io/configuration/configuration.py:
```diff
@@ -249,6 +271,35 @@ def set_optimization_definition(self, optimization_definition: Dict):
+                prob.driver = driver_instance
+            else:
+                # Raise error if no driver instance is provided
+                raise FASTConfigurationBaseKeyBuildingError(
+                    ValueError("No driver instance provided in configuration file"),
+                    KEY_DRIVER,
+                )
```
I think this code can be deleted, because the JSON schema already does this check.

ScottDelbecq replied:
Corrected in [3c7400f](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/3c7400f9d41f4d25e8c0a2c25ea1db12d5fa8941)

christophe-david commented on src/fastoad/io/configuration/resources/configuration.json:
```diff
@@ -134,9 +134,43 @@
+            }
+          },
+          "required": ["instance"],
+          "additionalProperties": false
```
You need to allow additional properties to have the `advanced_options` working as in the documentation.
```suggestion
          "additionalProperties": true
```

ScottDelbecq replied:
Corrected in [b907e10](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/b907e10574bbea0ccc9570d05088781856e0e8cd)

christophe-david commented on src/fastoad/io/configuration/tests/test_configuration.py:
```diff
@@ -336,3 +341,108 @@ def test_set_optimization_definition(cleanup):
+        "driver": {
+            "instance": "om.ScipyOptimizeDriver(optimizer='COBYLA')",
+            "options": {"maxiter": 100, "tol": 1e-2},
+        },
```
You miss a custom driver attribute here, like `advanced_options`.

ScottDelbecq replied:
Corrected [415aeec](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/415aeeca25f2e07ead12e0195b36b10a5992f7d3)

christophe-david commented on src/fastoad/io/configuration/configuration.py:
```diff
@@ -249,6 +271,35 @@ def set_optimization_definition(self, optimization_definition: Dict):
+                )
+
+            # Iterate over all keys (attributes) in driver_config except for 'instance'
+            for first_key, first_value in driver_config.items():
+                if first_key != "instance":
+                    # Iterate over all keys (values) in first_value
+                    for second_key, second_value in first_value.items():
+                        # driver.<first_key>[<second_key>] = second_value
+                        getattr(prob.driver, first_key)[second_key] = second_value
+
```
This would do the same job:
```suggestion
            for key, value in driver_config.items():
                if key != "instance":
                    getattr(prob.driver, key).update(value)
```

christophe-david commented on docs/documentation/usage.rst:
````diff
@@ -316,6 +285,108 @@ Several constraint variables can be defined.
+```
+
+### New Syntax
+In the new syntax, the driver is specified with an least an instance and more fields such as `options`.
````
```suggestion
In the new syntax, the driver is specified with at least an instance and more dict fields such as `options`.
```
(And maybe it's worth insisting even more on the fact that expected attributes are dictionaries)

ScottDelbecq replied:
Corrected in [d628427](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/d628427508cd3aaaf7da94b5589df728a6c7705f)

christophe-david reviewed (changes requested):
Thanks. This will be a great addition. 

I even think that the "advanced syntax" would be appreciated also for solver settings (but let's keep that for later).

Yet we still need some more adjustments.

ScottDelbecq commented on src/fastoad/io/configuration/configuration.py:
```diff
@@ -163,6 +168,23 @@ def load(self, conf_file: Union[str, PathLike]):
+        sys_paths = self._data.get("sys_paths", [])
+        for path in sys_paths:
+            if path not in sys.path:
+                sys.path.append(path)
```
Corrected in [e642cf0](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/e642cf0954f5ce8f8bb9c733f6f1c32d75ad2212)

ScottDelbecq commented on src/fastoad/io/configuration/configuration.py:
```diff
@@ -163,6 +168,23 @@ def load(self, conf_file: Union[str, PathLike]):
+            except (ImportError, AttributeError) as e:
+                raise ImportError(
+                    f"Failed to import {class_name} from {module_name} in configuration file: {e}"
+                )
```
Corrected in [aed1c1e](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/aed1c1e5e4b4d60451095b223a5c7a2da38ca228)

ScottDelbecq commented on docs/documentation/usage.rst:
```diff
@@ -316,6 +285,108 @@ Several constraint variables can be defined.
+      - /path/to/your/module
+
+This section is used to add a path to the Python sys.path.
+This is useful when you have a module that is not in the Python path.
```
Arf you are right there is a risk for confusion. I corrected in [1526415](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/1526415765f8b40ddd5624d6d4a0e62ea6d2a2aa) let me know if its clearer now.

christophe-david commented on docs/documentation/usage.rst:
```diff
@@ -316,6 +285,108 @@ Several constraint variables can be defined.
+
+The `driver` configuration can be specified in two ways: using the old syntax (a string) or the new syntax (an object).
+
+### Old Syntax
```
Far better :), but you really wanted "### Standard Syntax" and "### Advanced Syntax" to look like that?

ScottDelbecq replied:
I definitely want to continue mixing Mardown and reStructuredText 😅  
Should be good now in [479fa64](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/commits/479fa6428f2b802b0b118348f0e5af1771d5cd25)

christophe-david commented on docs/documentation/usage.rst:
```diff
@@ -316,6 +285,112 @@ Several constraint variables can be defined.
+
+In the standard syntax, the driver is specified as a string. This string should be a valid Python expression that creates an instance of an OpenMDAO driver.
+
+.. code:: yaml
+    driver: om.ScipyOptimizeDriver(tol=1e-2, optimizer='COBYLA')
```
Ok, but still not ok. This part does not appear in the rendered doc (I know, ReST sucks...)
```suggestion
.. code:: yaml

    driver: om.ScipyOptimizeDriver(tol=1e-2, optimizer='COBYLA')
```

christophe-david commented on docs/documentation/usage.rst:
```diff
@@ -316,6 +285,113 @@ Several constraint variables can be defined.
 Also, see :ref:`get-variable-list`.
 
+***************************
+Advanced configuration file
```
I just realized this:
```suggestion
Advanced configuration
```
I think this is important, because I don't want any user to be wondering if there is another "advanced" configuration file out there.

ScottDelbecq commented:
Thanks @christophe-david for the [modifications and additions](https://github.com/fast-aircraft-design/FAST-OAD/pull/559/files/6d7f7a187a1f8851071ef60c049624b127b0eea6..102e0e8cf47a0b36d7c108d431b4817091ffd1ff) it is OK for me!
